From 51d34e06242a9485b52d8922763747bf1e9cf54e Mon Sep 17 00:00:00 2001
From: Igor Opaniuk <igor.opaniuk@foundries.io>
Date: Tue, 2 Nov 2021 19:18:24 +0200
Subject: [PATCH 6/9] [FIO toup] zynqmp: add dynamic detection of fit image spi
 offset

Add dynamic detection of FIT image offset for QSPI boot.

Introduce SYS_SPI_BOOT_IMAGE_OFFS/SYS_SPI_BOOT_IMAGE_OFFS2 Kconfig
options that provide possibility to adjust spi boot image layout.

Testing (regular boot from primary boot image):
U-Boot SPL 2021.07+xlnx+gc5d8da6b06 (Nov 02 2021 - 21:49:46 +0000)
PMUFW:	v1.1
Loading new PMUFW cfg obj (2168 bytes)
Silicon version:	3
EL Level:	EL3
Multiboot:	0
Trying to boot from SPI
SPL: Booting next image from 0x100000 SPI offset

Testing (booting secondary boot image):
U-Boot SPL 2021.07+xlnx+g3f451431d6 (Nov 02 2021 - 22:09:42 +0000)
PMUFW:	v1.1
Loading new PMUFW cfg obj (2168 bytes)
Silicon version:	3
EL Level:	EL3
Multiboot:	10
Trying to boot from SPI
SPL: Booting next image from 0xaa0000 SPI offset

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>
Signed-off-by: Michael Scott <mike@foundries.io>
---
 arch/arm/mach-zynqmp/Kconfig         | 24 ++++++++++++++++++++++++
 board/xilinx/zynqmp/zynqmp.c         | 19 ++++++++++++++++---
 configs/xilinx_zynqmp_virt_defconfig |  1 +
 3 files changed, 41 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-zynqmp/Kconfig b/arch/arm/mach-zynqmp/Kconfig
index 46c7fc75537..5a782fb81c4 100644
--- a/arch/arm/mach-zynqmp/Kconfig
+++ b/arch/arm/mach-zynqmp/Kconfig
@@ -209,6 +209,30 @@ config CMD_ZYNQMP
 	  and authentication feature enabled while generating
 	  BOOT.BIN using Xilinx bootgen tool.
 
+config SYS_SPI_BOOT_IMAGE_OFFS
+	default 0x0
+	hex "address of primary boot image in SPI flash"
+	depends on SPL_SPI_LOAD
+	help
+	  Address within SPI-Flash from where the primary boot image
+	  is fetched from by BootROM. Must be a multiple of 32Kb (0x8000).
+
+config SYS_SPI_BOOT_IMAGE_OFFS2
+	default 0x50000
+	hex "address of secondary boot image in SPI flash"
+	depends on SPL_SPI_LOAD
+	help
+	  Address within SPI-Flash from where the secondary boot image
+	  is fetched from by BootROM. Must be a multiple of 32Kb (0x8000).
+
+config SYS_SPI_U_BOOT_OFFS2
+	default 0xaa0000
+	hex "address of secondary u-boot payload in SPI flash"
+	depends on SPL_SPI_LOAD
+	help
+	  Address within SPI-Flash from where the u-boot secondary payload
+	  is fetched from.
+
 source "board/xilinx/Kconfig"
 
 endif
diff --git a/board/xilinx/zynqmp/zynqmp.c b/board/xilinx/zynqmp/zynqmp.c
index aa9c7224605..1082e1d8d35 100644
--- a/board/xilinx/zynqmp/zynqmp.c
+++ b/board/xilinx/zynqmp/zynqmp.c
@@ -868,12 +868,25 @@ void set_dfu_alt_info(char *interface, char *devstr)
 unsigned int spl_spi_get_uboot_offs(struct spi_flash *flash)
 {
 	u32 offset;
+	u32 boot_image_offset = 0x0;
 	int multiboot = multi_boot_get();
 
-	offset = multiboot * SZ_32K;
-	offset += CONFIG_SYS_SPI_U_BOOT_OFFS;
+	boot_image_offset = golden_image_boundary * multiboot;
 
-	log_info("SPI offset:\t0x%x\n", offset);
+	/*
+	 * Default values are:
+	 * Primary boot.bin offset   - 0x0 (multiboot == 0)
+	 * Secondary boot.bin offset - 0x50000 (multiboot == 10,
+	 *                             as 10 * 32KB == 0x50000)
+	 */
+
+	if (boot_image_offset == CONFIG_SYS_SPI_BOOT_IMAGE_OFFS2) {
+		offset = CONFIG_SYS_SPI_U_BOOT_OFFS2;
+	} else {
+		offset = CONFIG_SYS_SPI_U_BOOT_OFFS;
+	}
+
+	printf("SPL: Booting next image from 0x%x SPI offset\n", offset);
 
 	return offset;
 }
diff --git a/configs/xilinx_zynqmp_virt_defconfig b/configs/xilinx_zynqmp_virt_defconfig
index cc298b200a1..d7e73295504 100644
--- a/configs/xilinx_zynqmp_virt_defconfig
+++ b/configs/xilinx_zynqmp_virt_defconfig
@@ -48,6 +48,7 @@ CONFIG_SPL_RAM_SUPPORT=y
 CONFIG_SPL_RAM_DEVICE=y
 CONFIG_SPL_SPI_LOAD=y
 CONFIG_SYS_SPI_U_BOOT_OFFS=0x100000
+CONFIG_SYS_SPI_U_BOOT_OFFS2=0xaa0000
 CONFIG_SPL_ATF=y
 CONFIG_SPL_ATF_NO_PLATFORM_PARAM=y
 CONFIG_SYS_MAXARGS=64
-- 
2.48.1

